waiter/src/waiter/service_description.clj

```diff
@@ -894,34 +894,43 @@
-                      :username username})
-              service-preauthorized (and token-preauthorized (empty? service-description-based-on-headers))
-              service-authentication-disabled (and token-authentication-disabled (empty? service-description-based-on-headers))
-              stored-service-description? (fetch-core kv-store service-id)]
-          ; Validating is expensive, so avoid validating if we've validated before, relying on the fact
-          ; that we'll only store validated service descriptions
-          (when-not stored-service-description?
-            (validate service-description-builder core-service-description {:allow-missing-required-fields? false})
-            (validate service-description-builder service-description {:allow-missing-required-fields? false}))
-          {:core-service-description core-service-description
-           :on-the-fly? contains-waiter-header?
-           :service-authentication-disabled service-authentication-disabled
-           :service-description service-description
-           :service-id service-id
-           :service-preauthorized service-preauthorized
-           :source-tokens source-tokens})
-        (catch [:type :service-description-error] ex-data
-          (throw (ex-info (:message ex-data)
-                          (merge (error-message-map-fn passthrough-headers waiter-headers) ex-data)
-                          (:throwable &throw-context))))))))
+      (let [{:keys [core-service-description service-description service-id]}
+            (build service-description-builder user-service-description
+                   {:assoc-run-as-user-approved? assoc-run-as-user-approved?
+                    :defaults defaults
+                    :kv-store kv-store
+                    :metric-group-mappings metric-group-mappings
+                    :service-id-prefix service-id-prefix
+                    :username username})
+            service-preauthorized (and token-preauthorized (empty? service-description-based-on-headers))
+            service-authentication-disabled (and token-authentication-disabled (empty? service-description-based-on-headers))]
+        {:core-service-description core-service-description
+         :on-the-fly? contains-waiter-header?
+         :service-authentication-disabled service-authentication-disabled
+         :service-description service-description
+         :service-id service-id
+         :service-preauthorized service-preauthorized
+         :source-tokens source-tokens})))
+
+  (defn validate-service-description
+    "Returns nil if the provided descriptor contains a valid service description.
+     Else it returns an instance of Throwable that reflects the validation error."
+    [kv-store service-description-builder
+     {:keys [core-service-description passthrough-headers service-description service-id waiter-headers]}]
+    (sling/try+
+      (let [stored-service-description? (fetch-core kv-store service-id)]
```

Comment:
The `?` on the end to me implies it's a boolean, but it's not.

Reply:
Removed the `?`